### PR TITLE
`SourceUtils.findMatchingChild` should never return a folder

### LINKED ISSUE #5738: Disassembled `*.class` opened instead of `*.java` when folder with name of class exists in sources JAR
### Apache NetBeans version

Apache NetBeans 17

### What happened

When you have a Maven sources JAR present in the local repository, confirmed by `SourceForBinaryQuery` via `apisupport.projectinspector`¹, then visiting a type or member in the classpath should open the editor on a read-only source file. In NB 16 this used to work consistently. In 17 it sometimes works, but sometimes you get the disassembled bytecode instead. I am still working out the precise conditions for the bug, since it happens on some types in a single JAR file and not others.

----

¹Sadly abandoned in `contrib` Hg rather than donated, AFAICT, but I still have a personal copy.

### How to reproduce

```bash
git clone https://github.com/jenkinsci/log-cli-plugin
cd log-cli-plugin
mvn -Pquick-build package dependency:sources
```

Now open this project in NB and wait for classpath scanning to complete. At https://github.com/jenkinsci/log-cli-plugin/blob/3d86457a9f2b19567d90542074224416a9ca474c/src/main/java/org/jenkinsci/plugins/log_cli/TailLogCommand.java#L55 place the caret on the call to `get` and press Ctrl-B to visit the definition.

Expected: you should see https://github.com/jenkinsci/jenkins/blob/jenkins-2.361.4/core/src/main/java/jenkins/model/Jenkins.java#L810

Actual: you will see disassembled bytecode

```java
    @NonNull
    public static Jenkins get() throws IllegalStateException {
        // <editor-fold defaultstate="collapsed" desc="Compiled Code">
        /* 0: invokestatic  #8                  // Method getInstanceOrNull:()Ljenkins/model/Jenkins;
         * 3: astore_0
         * 4: aload_0
         * 5: ifnonnull     18
         * 8: new           #9                  // class java/lang/IllegalStateException
         * 11: dup
         * 12: ldc           #10                 // String Jenkins.instance is missing. Read the documentation of Jenkins.getInstanceOrNull to see what you are doing wrong.
         * 14: invokespecial #11                 // Method java/lang/IllegalStateException."<init>":(Ljava/lang/String;)V
         * 17: athrow
         * 18: aload_0
         * 19: areturn
         *  */
        // </editor-fold>
    }
```

Note that if on the same line of source you instead click on the next method call, `checkPermission`, you are correctly taken to https://github.com/jenkinsci/jenkins/blob/jenkins-2.361.4/core/src/main/java/hudson/security/AccessControlled.java#L48 from `~/.m2/repository/org/jenkins-ci/main/jenkins-core/2.361.4/jenkins-core-2.361.4-sources.jar`.

### Did this work correctly in an earlier version?

Apache NetBeans 16

### Operating System

Xubuntu 22.04.2

### JDK

11.0.18; OpenJDK 64-Bit Server VM 11.0.18+10-post-Ubuntu-0ubuntu122.04; Ubuntu

### Apache NetBeans packaging

Apache NetBeans Snap Package

### Anything else

Using `git bisect` I was able to identify #5152 (@jtulach) as the cause of the regression. This does _not_ appear to be a duplicate of #5266, which apparently involved a visible exception (no error is printed in this case) and which was allegedly fixed in NB 17 in #5279 (@matthiasblaesing).

I hope to find time to study the bug and offer a PR, since this affects everyday usage of the IDE for me and is the one noticeable regression in NB 17.

### Are you willing to submit a pull request?

Yes

### PR DESCRIPTION
**Pull request by jglick**
Fixes #5738, a regression introduced by #5152 in NB 17.

### CODE REVIEW

File: java/java.source.base/src/org/netbeans/api/java/source/SourceUtils.java
Comment (jglick) [inline java/java.source.base/src/org/netbeans/api/java/source/SourceUtils.java:632]: Otherwise test may randomly pass or fail depending on filesystem iteration order. Would not affect production use since a `*-sources.jar` would return elements in ZIP order (typically folders first).

File: java/java.source.base/src/org/netbeans/api/java/source/SourceUtils.java
Comment (jglick) [inline java/java.source.base/src/org/netbeans/api/java/source/SourceUtils.java:690]: The key fix.

File: java/java.source.base/src/org/netbeans/api/java/source/SourceUtils.java
Comment (jglick) [inline java/java.source.base/src/org/netbeans/api/java/source/SourceUtils.java:1136]: #5152 made this method type-safe but neglected to adjust the Javadoc accordingly.

File: java/java.source.base/test/unit/src/org/netbeans/api/java/source/SourceUtilsTest.java
Comment (jglick) [inline java/java.source.base/test/unit/src/org/netbeans/api/java/source/SourceUtilsTest.java:341]: Existing test was too weak—tolerated any result, even completely wrong.

File: java/java.source.base/src/org/netbeans/api/java/source/SourceUtils.java
Comment (neilcsmith-net) [reply java/java.source.base/src/org/netbeans/api/java/source/SourceUtils.java:690]: OK with me, although I wonder if checking for `isJava()` first might be the right way as this partly overlaps?  Could this whole method just be ...

```java
return isJava(fo) && match(fo.getName(), name);
```

**Comment by neilcsmith-net** [comment]
Needs the Java label to run the correct CI test, although at present the "secret" lock/unlock to trigger a workflow run hasn't worked!

File: java/java.source.base/src/org/netbeans/api/java/source/SourceUtils.java
Comment (jglick) [reply java/java.source.base/src/org/netbeans/api/java/source/SourceUtils.java:690]: Well that is what was added in #5152, presumably necessary for that fix.

**Comment by mbien** [comment]
@neilcsmith-net seems to have worked now? The other workflow might have needed a few more minutes to cancel and exit before the new one started.

I have to take another look at the paperwork job on a rainy day since this one doesn't seem to cancel properly, it always finishes without getting the memo.

File: java/java.source.base/src/org/netbeans/api/java/source/SourceUtils.java
Comment (neilcsmith-net) [reply java/java.source.base/src/org/netbeans/api/java/source/SourceUtils.java:690]: Ah, missed that fact!  Only a concern if there's a file without extension that matches then?  Probably less likely.

**Comment by neilcsmith-net** [comment]
@mbien I just see "Workflow runs completed with no jobs" on this one?!

**Comment by mbien** [comment]
@neilcsmith-net oh right! Sorry I looked at the wrong PR in the console. I have seen it before that PR syncs didn't trigger new builds or triggered them with some delay - maybe this one will build a bit later.

**Comment by mbien** [comment]
https://www.githubstatus.com/ github is having some issues ATM. Adding do-not-merge so that nobody merges accidentally too early.

File: java/java.source.base/src/org/netbeans/api/java/source/SourceUtils.java
Comment (jglick) [reply java/java.source.base/src/org/netbeans/api/java/source/SourceUtils.java:690]: I suppose so. My limited understanding of the original change was that `name` could sometimes include the extension, like `X.scala` or something? At any rate, clearly it was intended to match on files and never folders.

**Review by mbien — APPROVED** [review]
makes sense to me, thanks for fixing this!

**Review by jlahoda — APPROVED** [review]
Looks good to me, thanks!

**Review by matthiasblaesing — APPROVED** [review]
Looks sane to me.

**Comment by mbien** [comment]
@jglick want to squash and merge?

**Comment by jglick** [comment]
Though it seems I have write access, I am not up to date on the technical procedures and conventions surrounding this repository so I would prefer if someone more active merged PRs.

**Comment by neilcsmith-net** [comment]
@jglick OK, if you squash the commits and force push updates to your PR branch, someone else (ie. me or whoever gets there first) can handle the merge when the CI completes.  Thanks!

**Comment by jglick** [comment]
> squash the commits and force push updates to your PR branch

Can do, though I generally dislike doing that as it makes it harder to track history and manage reviews. Out of curiosity, is this some policy for the repo, and if so, why? Squash-merging the PR would result in a single trunk commit anyway, the purported justification: https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=74681408#SubmittingPullRequests-SquashingcommitsonaPullRequest

**Comment by neilcsmith-net** [comment]
See eg. #5617 Squash-merging through the GitHub UI often has issues, and squash-merging locally and pushing to master loses the link with the PR (which we need).  No perfect answer! :smile:

**Comment by mbien** [comment]
there is also no requirement to have a PR produce a single commit. Sometimes it is better to separate things which don't belong together in one commit. So you force push into the branch to produce the cleaned up history, then you merge as is.

I personally find it rarely useful to go through the commit history in PRs (unless it has been cleaned up). Commits like "implemented review requests" have mostly only temporary value to communicate with other reviewers while the review is going on. Once its is reviewed, all what counts are the resulting commits (or single commit as in this case).

**Comment by mbien** [comment]
also as last step before merge I look at this:
https://github.com/apache/netbeans/pull/5739.patch

if the commit headers, emails co-authors etc are all correct I merge, without any github magic in between which might lead to surprises.

**Pull request merged**
